libraries/botbuilder-ai/src/luisRecognizer.ts

```diff
@@ -89,21 +108,25 @@ export class LuisRecognizer implements Middleware {
         if (force || !cached) {
             const utterance = context.activity.text || '';
             return this.luisClient.getIntentsAndEntitiesV2(this.settings.appId, this.settings.subscriptionKey, utterance, this.settings.options)
-                .then((result : LuisResult) => {
+                .then((luisResult : LuisResult) => {
                     // Map results
                     const recognizerResult: LuisRecognizerResult = {
-                        text: result.query,
-                        intents: this.getIntents(result),
-                        entities: this.getEntitiesAndMetadata(result.entities, result.compositeEntities, this.settings.verbose)
+                        text: luisResult.query,
+                        intents: this.getIntents(luisResult),
+                        entities: this.getEntitiesAndMetadata(luisResult.entities, luisResult.compositeEntities, this.settings.verbose)
                     };
-                    
                     // Write to cache
-                    context.services.set(this.cacheKey, recognizerResult);
-                    return recognizerResult;
+                    context.services.set(this.cacheKey, {recognizerResult, luisResult});
+                    
+                    return this.emitTraceInfo(context, luisResult, recognizerResult).then(() => {
+                        return recognizerResult;
+                    });
                 });
     
         }
-        return Promise.resolve(cached);
+        return this.emitTraceInfo(context, cached.recognizerResult, cached.luisResult).then(() => {
```

Comment:
>this.emitTraceInfo(context, cached.recognizerResult, cached.luisResult) [](start = 15, length = 71)

Do you really want to send multiple of these per/turn?  I would have thought just once when the initial call to recognize occurs...

Reply:
@Stevenic yes, we want all responses to include the trace activity so that the developer can inspect each one of them in the emulator. Otherwise we'll need to cache them in the emulator and figure out how to correlate the first call to future identical calls (i.e. your cache key logic).

Reply:
But that's not what that's doing... That's going to send a trace message every time the developer looks up the cached results for the received request. That means I could send 1 message to the bot and get 3 or 4 trace activities back depending on how my bot is coded. You're assuming the developer is only going to call `recognize()` once per/turn and the results are being cached so they can call it as many times as they want.

Reply:
Got it - thanks for explaining, I misunderstood the purpose of the cache and though it was to reduce LUIS service calls across turns, I re-read the code and now your original feedback makes more sense. 

I changed it so that it wouldn't emit the trace activity on cache-hits and I updated the unit tests to ensure that the activity is emitted only once on multiple calls to recognize() on the same context.